### PR TITLE
Port overview

### PR DESCRIPTION
**Pull request by attiks**
- [x] Code ported
- [x] Code linted
- [x] Tested

Refs: #RW-606

### CODE REVIEW

File: src/overview.js [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+ * Overview board manager.
+ */
+function OverviewManager(config, logger, trelloClient, date) {
+  this.config = config;
+  this.logger = logger;
+  this.trelloClient = trelloClient;
+
+  this.overviewBoard = null;
+  this.boards = null;
+  this.projects = null;
```
Comment (orakili) [inline src/overview.js]: This should be initialized to `{}` or `new Map()`. (The latter may ease iteration and lookup).

File: src/overview.js [hunk 2]
```diff
@@ -0,0 +1,360 @@
+/**
+ * Overview board manager.
+ */
+function OverviewManager(config, logger, trelloClient, date) {
+  this.config = config;
+  this.logger = logger;
+  this.trelloClient = trelloClient;
+
+  this.overviewBoard = null;
+  this.boards = null;
```
Comment (orakili) [inline src/overview.js]: This should be initialized to `[]`.

File: src/overview.js [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+  this.projects = null;
+
+  /**
+   * Main process function.
+   */
+  this.process = async () => {
+    try {
+      if (this.getBoards()) {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      if (await this.getBoards()) {
```

File: src/overview.js [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    catch (exception) {
+      this.logger.error('Unable to load board ' + board.id);
+    }
+  };
+
+  /**
+   * Retrieve organization boards.
+   */
+  this.getBoards = async () => {
+    const boards = await this.trelloClient.get('/organizations/' + this.config.organization + '/boards', {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
    const boards = await this.trelloClient.get('/organizations/' + this.config.trello.organization + '/boards', {
```

File: src/overview.js [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+      filter: 'open',
+      fields: 'name,shortUrl,labels,idList,due',
+    };
+
+    if (checklists) {
+      date['checklists'] = 'all';
+    }
+
+    try {
+      board.cards = await this.trelloClient.get('/boards/' + board.id, data);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      board.cards = await this.trelloClient.get('/boards/' + board.id + '/cards', data);
```

File: src/overview.js [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+   * Retrieve board data.
+   */
+  this.getBoardData = async (board, checklists) => {
+    const data = {
+      filter: 'open',
+      fields: 'name,shortUrl,labels,idList,due',
+    };
+
+    if (checklists) {
+      date['checklists'] = 'all';
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      data['checklists'] = 'all';
```

File: src/overview.js [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+   */
+  this.getBoards = async () => {
+    const boards = await this.trelloClient.get('/organizations/' + this.config.organization + '/boards', {
+      filter: 'open',
+      fields: 'id,name',
+      lists: 'open',
+    });
+
+    for (let board of boards) {
+      if (board.id == this.config.boardId) {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      if (board.id == this.config.trello.boardId) {
```

File: src/overview.js [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+        this.getBoardData(board, false);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
        await this.getBoardData(board, true);
```

File: src/overview.js [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+        this.overviewBoard = board;
+      }
+      else {
+        this.getBoardData(board, false);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
        await this.getBoardData(board, false);
```

File: src/overview.js [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+  this.update = async () => {
+    let boards = new Map();
+    for (const board of this.boards) {
+      boards.set(board.name, true);
+    }
+
+    for (const project of this.projects) {
+      let card = project.card;
+      for (const checklist of card.checklist) {
+        if (boards[checklist.name]) {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
        if (boards.has(checklist.name)) {
```

File: src/overview.js [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+          let actions = project.actions[checklist.name];
```
Comment (orakili) [inline src/overview.js]:
```suggestion
          let actions = project.actions.get(checklist.name);
```

File: src/overview.js [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+          if (actions) {
+            this.updateChecklist(card.id, checklist, actions);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
            await this.updateChecklist(card.id, checklist, actions);
```

File: src/overview.js [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+            project.actions.delete(checklist.name);
+          }
+          else {
+            this.deleteChecklist(checklist.id);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
            await this.deleteChecklist(checklist.id);
```

File: src/overview.js [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+          }
+        }
+        else {
+          this.deleteChecklist(checklist.id);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
          await this.deleteChecklist(checklist.id);
```

File: src/overview.js [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+        }
+      }
+
+      for (const [checklistName, actions] of project.actions) {
+        this.addChecklist(card.id, checklistName, actions);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
        await this.addChecklist(card.id, checklistName, actions);
```

File: src/overview.js [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+      items.set(action.link, action);
+    }
+
+    for (const checkitem of checklist.checkItems) {
+      if (checkitem.name.length > 29) {
+        checkitem.name = checkitem.name.slice(0, 29);
+        if (items.has(checkitem.name)) {
+          let action = items.get(checkitem.name);
+          if (action.name != checkitem.name) {
+            this.updateCheckItem(cardId, checklist.id, checkitem.id, action);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
           await this.updateCheckItem(cardId, checklist.id, checkitem.id, action);
```

File: src/overview.js [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+            items.delete(checkitem.name);
+            continue;
+          }
+        }
+      }
+      else {
+        this.deleteCheckItem(checklist.id, checkitem.id);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
        await this.deleteCheckItem(checklist.id, checkitem.id);
```

File: src/overview.js [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+      }
+    }
+
+    for (const action of items) {
+      this.addCheckItem(checklist.id, action);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      await this.addCheckItem(checklist.id, action);
```

File: src/overview.js [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+  this.addChecklist = async (cardId, name, actions) => {
+    try {
+      const checklistId = await this.trelloClient.post('/checklists', {
+        idCart:  cardId,
+        name:  name,
+      });
+      this.logger.debug('Added checklist ' + name);
+
+      for (const action of actions) {
+        this.addCheckItem(checklistId, action);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
        await this.addCheckItem(checklistId, action);
```

File: src/overview.js [hunk 20]
```diff
@@ -0,0 +1,360 @@
...
+    }
+
+    return this.projects;
+  };
+
+  /**
+   * Prepare the data to update the overview board.
+   */
+  this.prepare = async () => {
+    let projects = this.projects;
```
Comment (orakili) [inline src/overview.js]:
```suggestion
    let projects = this.getProjects();
```

File: src/overview.js [hunk 21]
```diff
@@ -0,0 +1,360 @@
...
+      let lists = {};
+      for (const list of board.lists) {
+        lists[list.id] = list.name;
+      }
+
+      for (const card of board.card) {
+        let actions = this.getActions(card, lists[card.idList]);
+        for (const [name, action] of actions) {
+          if (projects[name]) {
+            if (!projects.actions[board.name]) {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
            if (!projects.actions.has(board.name)) {
```

File: src/overview.js [hunk 22]
```diff
@@ -0,0 +1,360 @@
...
+              projects.actions[board.name] = [];
```
Comment (orakili) [inline src/overview.js]:
```suggestion
              projects.actions.set(board.name, []);
```

File: src/overview.js [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+            }
+            projects.actions[board.name].push(action);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
            projects.actions.get(board.name).push(action);
```

File: src/overview.js [hunk 24]
```diff
@@ -0,0 +1,360 @@
...
+   * Get Overview board projects.
+   */
+  this.getProjects = async () => {
+    for (const card of this.overviewBoard.cards) {
+      for (const label of card.labels) {
+        let projectName = this.getProjectName(label.name);
+        if (projectName != '') {
+          this.projects[projectName] = {
+            card: card,
+            actions: []
```
Comment (orakili) [inline src/overview.js]:
```suggestion
            actions: new Map(),
```

File: src/overview.js [hunk 25]
```diff
@@ -0,0 +1,360 @@
...
+      }
+    }
+
+    return actions;
+  };
+
+  /**
+   * Get Overview board projects.
+   */
+  this.getProjects = async () => {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
  this.getProjects = () => {
```

File: src/overview.js [hunk 26]
```diff
@@ -0,0 +1,360 @@
...
+      }
+    }
+
+    return this.projects;
+  };
+
+  /**
+   * Prepare the data to update the overview board.
+   */
+  this.prepare = async () => {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
  this.prepare = () => {
```

File: src/overview.js [hunk 27]
```diff
@@ -0,0 +1,360 @@
...
+      parts.push(parts, '*' + action.Due + '*');
+    }
+
+    return parts.join(' - ');
+  };
+
+  /**
+   * Get the projects actions for a card.
+   */
+  this.getActions = async (card, list) => {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
  this.getActions = (card, list) => {
```

File: src/overview.js [hunk 28]
```diff
@@ -0,0 +1,360 @@
...
+
+    for (const label of card.labels) {
+      let projectName = this.getProjectName(label.name);
+      if (projectName != '') {
+        actions[projectName] = {
+          link: card.ShortUrl,
+          status: list,
+          due: card.Due,
+          complete: false,
+        };
```
Comment (orakili) [inline src/overview.js]:
```suggestion
        actions.set(projectName, {
          link: card.ShortUrl,
          status: list,
          due: card.Due,
          complete: false,
        });
```

File: src/overview.js [hunk 29]
```diff
@@ -0,0 +1,360 @@
...
+      return label.substring(this.config.doerPrefix.length);
+    }
+
+    return '';
+  };
+
+  /**
+   * Compute an action name.
+   */
+  this.getActionName = async (action) => {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
  this.getActionName = (action) => {
```

File: src/overview.js [hunk 30]
```diff
@@ -0,0 +1,360 @@
...
+    let doers = '*Not assigned*';
+    if (action.Doers.length > 0) {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
    if (action.doers.length > 0) {
```

File: src/overview.js [hunk 31]
```diff
@@ -0,0 +1,360 @@
...
+      doers = '**' + action.Doers.join(', ') + '**';
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      doers = '**' + action.doers.join(', ') + '**';
```

File: src/overview.js [hunk 32]
```diff
@@ -0,0 +1,360 @@
...
+    }
+
+    let parts = [
+      action.Link,
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      action.link,
```

File: src/overview.js [hunk 33]
```diff
@@ -0,0 +1,360 @@
...
+      doers,
+      action.Status
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      action.status
```

File: src/overview.js [hunk 34]
```diff
@@ -0,0 +1,360 @@
...
+    ];
+
+    if (action.Due != '') {
```
Comment (orakili) [inline src/overview.js]:
```suggestion
    if (action.due != '') {
```

File: src/overview.js [hunk 35]
```diff
@@ -0,0 +1,360 @@
...
+      parts.push(parts, '*' + action.Due + '*');
```
Comment (orakili) [inline src/overview.js]:
```suggestion
      parts.push(parts, '*' + action.due + '*');
```

File: src/overview.js [hunk 36]
```diff
@@ -0,0 +1,360 @@
...
+
+    return '';
+  };
+
+  /**
+   * Get doer name.
+   */
+  this.getDoerName = async (label) => {
+    if (label.indexOf(this.config.doerPrefix) === 0) {
+      return label.substring(this.config.doerPrefix.length);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
    if (label.indexOf(this.config.trello.doerPrefix) === 0) {
      return label.substring(this.config.trello.doerPrefix.length);
```

File: src/overview.js [hunk 37]
```diff
@@ -0,0 +1,360 @@
...
+      this.logger.error(exception);
+    }
+  };
+
+  /**
+   * Get project name.
+   */
+  this.getProjectName = async (label) => {
+    if (label.indexOf(this.config.projectPrefix) === 0) {
+      return label.substring(this.config.projectPrefix.length);
```
Comment (orakili) [inline src/overview.js]:
```suggestion
    if (label.indexOf(this.config.trello.projectPrefix) === 0) {
      return label.substring(this.config.trello.projectPrefix.length);
```

File: src/overview.js [hunk 38]
```diff
@@ -0,0 +1,360 @@
...
+  this.boards = null;
+  this.projects = null;
+
+  /**
+   * Main process function.
+   */
+  this.process = async () => {
+    try {
+      if (this.getBoards()) {
+        await this.prepare();
```
Comment (orakili) [inline src/overview.js]:
```suggestion
        this.prepare();
```

**Review by orakili — COMMENTED** [review]
Sorry for the noise, just went through the code and marked what jumped at me. Probably missed some and some suggestions may need to be confirmed.

This app is indeed difficult to test. Maybe the setup could be reproduced by creating 2-3 private boards in your account, one being the "overview" board and then creating lists and cards with the `Doer: XXX` and `Project: YYY` labels.

**Comment by attiks** [comment]
- Working code
- Each project can have a maximum of 1 card?
- ~~Removing the Id from the `for ... of` loops breaks the code~~

**Comment by orakili** [comment]
> * Each project can have a maximum of 1 card?

Yes, normally (i.e if people follow the instructions correctly), there's supposed to be only 1 card with a given `Project: YYY` label on the overview board.

File: src/overview.js [hunk 39]
```diff
@@ -0,0 +1,385 @@
...
+  };
+
+  /**
+   * Update board checklist: add, update or remove actions.
+   */
+  this.updateChecklist = async (cardId, checklist, actions) => {
+    let items = new Map();
+    for (const action of actions.values()) {
+      this.logger.debug('Processing action ' + action.name);
+      items.set(action.name, action);
```
Comment (orakili) [inline src/overview.js]: This should use the `action.link` as key as it's the only non changing part of the check list items ("link - does - status - due"). See next comment below.

File: src/overview.js [hunk 40]
```diff
@@ -0,0 +1,385 @@
...
+    }
+
+    for (const checkitem of checklist.checkItems) {
+      if (items.has(checkitem.name)) {
```
Comment (orakili) [inline src/overview.js]: This should extract the link from the check list item name ("link - does - status - due") to do the lookup in the `items` map.

`action.name` may indeed differ from `checkitem.name` (previous value), for example if the status, doer or due changed. Only the beginning with the link is the same and serves as identifier.

Ref: https://github.com/UN-OCHA/legacy--rw-trello-overview/blob/fe4a507e53bf6d3fc28c58c5d3c064e5020bfa65/main.go#L231-L238

**Review by orakili — CHANGES_REQUESTED** [review]
Looks good but the existing check list items are not updated properly when something changed like doer, status (card list) or due date.

File: src/overview.js (hunk 39 above)
Comment (attiks) [reply src/overview.js]: Changed

File: src/overview.js (hunk 40 above)
Comment (attiks) [reply src/overview.js]: Changed, but :100: % not sure

**Review by orakili — APPROVED** [review]

**Pull request merged**